server/src/main/java/org/elasticsearch/index/mapper/JsonFieldParser.java

```diff
@@ -31,53 +31,117 @@
 
 /**
  * A helper class for {@link JsonFieldMapper} parses a JSON object
- * and produces an indexable field for each leaf value.
+ * and produces a pair of indexable fields for each leaf value.
  */
 public class JsonFieldParser {
+    private static final String SEPARATOR = "\0";
```

Comment:
The null character `\0` seemed like a reasonable choice for a separator, as (1) it shouldn’t show up too often in field keys, and (2) there is already precedent for it, as we use it when storing percolator queries (`PercolatorFieldMapper#FIELD_VALUE_SEPARATOR`).